server/src/main/java/org/gluu/oxtrust/action/Authenticator.java

```diff
@@ -193,33 +217,34 @@ private GluuCustomPerson findPersonByDn(String userDn) {
 
 	/**
 	 * Main entry point for oAuth authentication.
-	 * 
+	 *
 	 * @throws IOException
-	 * 
+	 *
 	 * @throws Exception
 	 */
 	public boolean oAuthLogin() throws IOException, Exception {
-		Client client = ClientBuilder.newClient();
-		WebTarget target = client.target(openIdService.getOpenIdConfiguration().getAuthorizationEndpoint());
-		String clientId = appConfiguration.getOxAuthClientId();
-		String scope = appConfiguration.getOxAuthClientScope();
-		String responseType = "code";
+		GluuConfiguration configuration = configurationService
+				.getConfiguration(new String[] { "oxTrustAuthenticationMode" });
+		String acrValues = configuration.getOxTrustAuthenticationMode();
 		String nonce = UUID.randomUUID().toString();
 		String state = UUID.randomUUID().toString();
-		target = target.queryParam(OxTrustConstants.OXAUTH_CLIENT_ID, clientId);
-		target = target.queryParam(OxTrustConstants.OXAUTH_REDIRECT_URI, appConfiguration.getLoginRedirectUrl());
-		target = target.queryParam(OxTrustConstants.OXAUTH_RESPONSE_TYPE, responseType);
-		target = target.queryParam(OxTrustConstants.OXAUTH_SCOPE, scope);
-		target = target.queryParam(OxTrustConstants.OXAUTH_NONCE, nonce);
-		target = target.queryParam(OxTrustConstants.OXAUTH_STATE, state);
+		String clientId = appConfiguration.getOxAuthClientId();
+		String scope = appConfiguration.getOxAuthClientScope();
+
+		Client client = ClientBuilder.newClient();
+		WebTarget target = client.target(openIdService.getOpenIdConfiguration().getAuthorizationEndpoint());
+		if (this.fapiCompatibility) {
```

Comment:
@miltonbo oxauth's `fapiCompatibility` configuration flag defines oxauth behavior. It doesn't influence client authentication method. Means that oxauth can perfectly run with `fapiCompatibility=true` without MTLS. 
Client request should be build based on `client_id` configuration.

`clientService.getClientByInum(String inum).getTokenEndpointAuthMethod()` : 
- `tls_client_auth` or `self_signed_tls_client_auth` then MTLS
- `private_key_jwt` then private key.

Goal is to construct it based on client configuration, not globally for whole app.

Reply:
Ok, in general I understand that I have to read client configuration to do the authentication process. This was my doubt and I understood you that I have to put it in oxTrust json config.
Also I didn't be very sure to use clientService in authentication process, since in oxTrust json config we have the property `oxAuthClientPassword`, that value could be read directly using clientService also, but someone in the past put that value in oxTrust json config for some reason.

Reply:
Good question. I can't think about any reason why produce duplication. If client is not set to authenticate it by `tls_client_auth` then what would be the reason to send that request at all. re: `oxAuthClientPassword`, I think you can drop it since as you said you can read that property from persistence.

Reply:
Got it. My last but not least question, you said that we could use `fapiCompatibility=true` without MTLS, but I'm not really sure whether that will work, since we have Apache working over Jetty to process MTLS and it share us the header that we use to validate the client. Just to be sure whether I have to send the client certificate using `fapiCompatibility=true` from oxTrust in every rest call (token endpoint for instance).

Reply:
I meant that you shouldn't check `fapiCompatibility`. Instead you should check whether client's authentication method is `tls_client_auth` or `self_signed_tls_client_auth`. If yes, then send certificate, if not then does not send certificate. Request builder should be bound to authentication method. It easiest and clear way.

Reply:
Ok, I'll do that, thanks a lot.